File: backend/app/workers/jobs/notify.py
"""Feina programada: enviament dels avisos per correu.

Cada espai te els seus destinataris: l'avis d'un descobert a Calella nomes va
a qui li pertoca. Els avisos que no son de cap espai (connexions, sincronitza-
cions) van als destinataris generals de la configuracio.
"""

from __future__ import annotations

import logging
from collections import defaultdict

from sqlalchemy import select
from sqlalchemy.orm import Session

from app.config import settings
from app.core.time import today_local, utcnow
from app.db import session_scope
from app.models import Alert, Ledger
from app.models.enums import AlertSeverity, AlertStatus
from app.notifications.email import render_digest, send_email

logger = logging.getLogger(__name__)


def run_notification_job(only_critical: bool = False, db: Session | None = None) -> str:
    """Envia els avisos encara no notificats.

    Amb `only_critical` nomes surten els urgents, perque es pugui cridar cada
    hora sense omplir la bustia; la resta van al resum diari.
    """
    if db is not None:
        return notify_pending(db, only_critical)
    with session_scope() as session:
        return notify_pending(session, only_critical)


def recipients_for(ledger: Ledger | None) -> list[str]:
    """A qui van els avisos d'aquest espai."""
    if ledger is not None and ledger.alert_recipients:
        return list(ledger.alert_recipients)
    return list(settings.alert_recipients)


def notify_pending(db: Session, only_critical: bool = False) -> str:
    query = select(Alert).where(Alert.notified_at.is_(None), Alert.status != AlertStatus.DISMISSED)
    if only_critical:
        query = query.where(Alert.severity == AlertSeverity.CRITICAL)
    alerts = list(db.scalars(query.order_by(Alert.severity, Alert.created_at)))

    if not alerts:
        return "Cap avis pendent d'enviar"

    per_espai: dict[int | None, list[Alert]] = defaultdict(list)
    for alert in alerts:
        per_espai[alert.ledger_id].append(alert)

    enviats = 0
    pendents = 0
    for ledger_id, del_espai in per_espai.items():
        ledger = db.get(Ledger, ledger_id) if ledger_id is not None else None
        destinataris = recipients_for(ledger)
        if not destinataris:
            logger.info(
                "Sense destinataris per a %s: %s avisos queden pendents",
                ledger.name if ledger else "avisos generals",
                len(del_espai),
            )
            pendents += len(del_espai)
            continue

        if only_critical:
            title = "Avis urgent de la comptabilitat"
            subtitle = "Hi ha una cosa que necessita atencio ara."
        else:
            title = "Resum d'avisos"
            subtitle = f"Avisos nous del {today_local():%d/%m/%Y}."
        if ledger is not None:
            subtitle = f"{ledger.name} · {subtitle}"

        html, text = render_digest(del_espai, title, subtitle)
        nom = f"{title} · {ledger.name}" if ledger else title
        subject = (
            f"{nom}: {del_espai[0].title}" if len(del_espai) == 1 else f"{nom} ({len(del_espai)})"
        )

        if not send_email(subject, html, text, recipients=destinataris):
            pendents += len(del_espai)
            continue

        sent_at = utcnow()
        for alert in del_espai:
            alert.notified_at = sent_at
        enviats += len(del_espai)

    db.flush()

    if enviats and pendents:
        return f"{enviats} avisos enviats; {pendents} pendents (sense destinatari o error)"
    if enviats:
        return f"{enviats} avisos enviats per correu"
    return f"{pendents} avisos pendents: no hi ha destinataris o el correu ha fallat"

File: src/services/notify.ts
/**
 * Sending the alerts by email.
 *
 * Each workspace has its own recipients: the alert about an overdraft in
 * Calella only goes to whoever it concerns. Alerts belonging to no workspace
 * (connections, synchronizations) go to the general recipients in the
 * configuration.
 *
 * A translation of `backend/app/workers/jobs/notify.py`.
 */

import { and, asc, eq, isNull, ne } from "drizzle-orm";

import { db } from "../db/client.ts";
import { alerts, ledgers, type Alert } from "../db/schema/index.ts";
import { config } from "../lib/config.ts";
import { sendMail, renderSummary, type SummaryEntry } from "../lib/email.ts";
import { todayLocal } from "../lib/time.ts";

/** Local date and time, as `strftime("%d/%m/%Y %H:%M")` wrote them. */
const localStamp = new Intl.DateTimeFormat("ca-ES", {
  day: "2-digit",
  month: "2-digit",
  year: "numeric",
  hour: "2-digit",
  minute: "2-digit",
  hour12: false,
  timeZone: config.timezone,
});

function formatStamp(when: Date): string {
  // The Catalan `Intl` puts «, » between the date and the time; the Python does not.
  return localStamp.format(when).replace(", ", " ");
}

function dateShort(isoDate: string): string {
  const [any, month, day] = isoDate.split("-");
  return `${day}/${month}/${any}`;
}

/** Who this workspace's alerts go to. */
export function recipientsOf(workspaceRecipients: readonly string[] | null): string[] {
  if (workspaceRecipients !== null && workspaceRecipients.length > 0)
    return [...workspaceRecipients];
  return [...config.alertRecipients];
}

/**
 * Sends the alerts that have not been notified yet.
 *
 * With `nomesUrgents` only the critical ones come out, so that it can be
 * called every hour without filling the inbox; the rest go in the daily digest.
 */
export async function notifyPending(urgentOnly = false): Promise<string> {
  const conditions = [isNull(alerts.notifiedAt), ne(alerts.status, "dismissed")];
  if (urgentOnly) conditions.push(eq(alerts.severity, "critical"));

  const pending = await db
    .select()
    .from(alerts)
    .where(and(...conditions))
    // Same order as the Python: `severity` is text, so alphabetically
    // «critical» < «info» < «warning» and the urgent ones come first.
    .orderBy(asc(alerts.severity), asc(alerts.createdAt));

  if (pending.length === 0) return "Cap avis pendent d'enviar";

  const byWorkspace = new Map<number | null, Alert[]>();
  for (const alert of pending) {
    const key = alert.ledgerId;
    const list = byWorkspace.get(key);
    if (list === undefined) byWorkspace.set(key, [alert]);
    else list.push(alert);
  }

  let sent = 0;
  let pendingUnsent = 0;

  for (const [ledgerId, workspaceAlerts] of byWorkspace) {
    const [workspace] =
      ledgerId === null
        ? []
        : await db.select().from(ledgers).where(eq(ledgers.id, ledgerId)).limit(1);

    const recipients = recipientsOf(workspace?.alertRecipients ?? null);
    if (recipients.length === 0) {
      console.info(
        `[avisos] sense destinataris per a ${workspace?.name ?? "avisos generals"}: ` +
          `${workspaceAlerts.length} avisos queden pendents`,
      );
      pendingUnsent += workspaceAlerts.length;
      continue;
    }

    const title = urgentOnly ? "Avis urgent de la comptabilitat" : "Resum d'avisos";
    let subtitle = urgentOnly
      ? "Hi ha una cosa que necessita atencio ara."
      : `Avisos nous del ${dateShort(todayLocal())}.`;
    if (workspace !== undefined) subtitle = `${workspace.name} · ${subtitle}`;

    const entries: SummaryEntry[] = workspaceAlerts.map((alert) => ({
      severity: alert.severity,
      title: alert.title,
      body: alert.body,
      ledgerName: workspace?.name ?? "",
      created: formatStamp(alert.createdAt),
    }));

    const { html, text } = await renderSummary(entries, title, subtitle);
    const name = workspace !== undefined ? `${title} · ${workspace.name}` : title;
    const first = workspaceAlerts[0];
    const subject =
      workspaceAlerts.length === 1 && first !== undefined
        ? `${name}: ${first.title}`
        : `${name} (${workspaceAlerts.length})`;

    if (!(await sendMail(subject, html, text, recipients))) {
      pendingUnsent += workspaceAlerts.length;
      continue;
    }

    const now = new Date();
    for (const alert of workspaceAlerts) {
      await db.update(alerts).set({ notifiedAt: now }).where(eq(alerts.id, alert.id));
    }
    sent += workspaceAlerts.length;
  }

  if (sent > 0 && pendingUnsent > 0) {
    return `${sent} avisos enviats; ${pendingUnsent} pendents (sense destinatari o error)`;
  }
  if (sent > 0) return `${sent} avisos enviats per correu`;
  return `${pendingUnsent} avisos pendents: no hi ha destinataris o el correu ha fallat`;
}
